### PR TITLE
MSC xxxx: Resolve URL API

### PR DESCRIPTION
**Pull request by uhoreg**
This is a markdown-ification of https://github.com/matrix-org/matrix-doc/issues/1270, with some clarifications and expansions added.

### CODE REVIEW

File: proposals/2370-resolve-url.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+## Proposal
+
+Add a new endpoint to resolve a URL.
+
+Example:
+
+Request:
+```
+POST /_matrix/media/v1/resolve_url
```
Comment (uhoreg) [inline proposals/2370-resolve-url.md]: Would this be better as a `GET` request (like `GET /_matrix/media/r0/preview_url` )?

File: proposals/2370-resolve-url.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+Errors can occur when processing the request.
+
+* If the resource cannot be downloaded due to the URL returning a 404 error, the
+  endpoint should respond with an `M_NOT_FOUND` error and HTTP code 404.
+* If the resource cannot be downloaded due to the URL returning another error,
+  the endpoint should respond with an `M_UNKNOWN` error and HTTP code 400.
+* If the resource cannot be downloaded because it has been blacklisted, it should
+  respond with an `M_FORBIDDEN` error and HTTP code 403.
+* If the resource is larger than the homeserver's upload size limit, it should
+  respond with an `M_TOO_LARGE` error and HTTP code 400.
```
Comment (uhoreg) [inline proposals/2370-resolve-url.md]: The original proposal listed this an HTTP code 413.  However it seems like 413 is for when the request payload is too large, rather than for when a referenced resource is too large, so I think that just a generic 400 would be better here.

File: proposals/2370-resolve-url.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+and return the `mxc:` URL for the resource, along with the MIME type.
+
+Errors can occur when processing the request.
+
+* If the resource cannot be downloaded due to the URL returning a 404 error, the
+  endpoint should respond with an `M_NOT_FOUND` error and HTTP code 404.
+* If the resource cannot be downloaded due to the URL returning another error,
+  the endpoint should respond with an `M_UNKNOWN` error and HTTP code 400.
+* If the resource cannot be downloaded because it has been blacklisted, it should
+  respond with an `M_FORBIDDEN` error and HTTP code 403.
```
Comment (uhoreg) [inline proposals/2370-resolve-url.md]: For reference, synapse returns `M_UNKNOWN` with HTTP code 403 on `/preview_url`, but using `M_FORBIDDEN` seems more useful.

File: proposals/2370-resolve-url.md (hunk 1 above)
Comment (anoadragon453) [reply proposals/2370-resolve-url.md]: Like `GET /_matrix/media/v1/resolve_url?url=https://blabla.com`? I was going to say it's annoying to have to URL-encode a URL, but seeing as `/preview_url` already requires this then it's probably fine.

Also why `v1` whereas `/preview_url` is `r0`?

File: proposals/2370-resolve-url.md (hunk 3 above)
Comment (anoadragon453) [reply proposals/2370-resolve-url.md]: Indeed, another MSC time? :)

File: proposals/2370-resolve-url.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+* Only authorized users should be allowed to call this endpoint.
+* This endpoint should be rate limited.
+* Homeservers may wish to restrict the MIME types that it is willing to
+  download (for example, only images).
+* Homeservers should allow administrators to blacklist URL patterns and IP
+  addresses to prevent users from being able to access resources that are only
+  available on an internal network.
+
+Since the homeserver downloads the file, this will not work for trying to
+upload encrypted files for end-to-end encrypted rooms.
```
Comment (anoadragon453) [inline proposals/2370-resolve-url.md]: In this case, when a matrix web-client gets a URL it could just download it itself? Or do we want to prevent requiring any outgoing connections from a client except for matrix-client.domain.com and matrix-server.domain.com?

File: proposals/2370-resolve-url.md (hunk 1 above)
Comment (uhoreg) [reply proposals/2370-resolve-url.md]: > Also why `v1` whereas `/preview_url` is `r0`?

Indeed.  I copied this from the original doc, but the rest of `/media` uses `r0`, so this should too.  Fixed.

As far as `GET` vs `POST`, I suppose it could be argued that since it causes something to be stored on the server, it should be a `POST`, whereas `/preview_url` does not cause anything to be stored on the server (beyond possible caching), so it's fine as a `GET`.  I don't really have strong opinions either way, so if anyone does have opinions, please say so.

Comment (turt2live) [reply proposals/2370-resolve-url.md]: Annoyingly media is weird. It's still in the v1 version.

File: proposals/2370-resolve-url.md (hunk 4 above)
Comment (uhoreg) [reply proposals/2370-resolve-url.md]: If it's a web-client, it may not be able to download it itself due to CORS.  I don't think we should be telling clients what they *should* do in this case, but a client could just refuse to attach the image and tell the user to attach it themselves (e.g. by downloading and uploading, or copy-and-paste).

File: proposals/2370-resolve-url.md (hunk 1 above)
Comment (uhoreg) [reply proposals/2370-resolve-url.md]: Spec says `r0` though. https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-media-r0-preview-url :-/

**Comment by NotAFile** [comment]
I don't quite understand why this is necessary. The URL preview API already returns images for links in messages, which riot displays, albeit smaller than full images:

image

The way some platforms like e.g. Telegram do this is to show message that contain image links with a preview that is very similar in size to posting a full image:

image

So the data is basically already there, Riot just needs to display image URL previews as big as images.

Doing this over the approach in this MSC has a few advantages in my opinion:

 1. No new API required. No additional Complexity is needed for servers and clients to implement.
 2. Saves storage Space. While the server still needs to proxy the image for privacy reasons, the server does not need to store the image forever. It can cache it short term and re-download it again later on demand.
 3. It requires no changes on the sending clients side, so it instantly works for everyone's messages. It also works in any message history predating this feature.
 4. It preserves the context that this is an image that's not created by the user, but downloaded from some URL.

**Comment by turt2live** [comment]
@NotAFile please use a comment on the diff so others can reply more easily, even if it means putting the comment onto the title.

File: proposals/2370-resolve-url.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
+# Resolve URL API
+
+(This document is heavily based on
+https://docs.google.com/document/d/1bbX1yxNETmMa-AxBGjIpb4lNoTuc3vjGXmbZWrNBlzM/edit)
+
+Users sometimes want to upload images to Matrix rooms that they see on other
```
Comment (NotAFile) [inline proposals/2370-resolve-url.md]: I don't quite understand why this is necessary. The URL preview API already returns images for links in messages, which riot displays, albeit smaller than full images:

image

The way some platforms like e.g. Telegram do this is to show message that contain image links with a preview that is very similar in size to posting a full image:

image

So the data is basically already there, Riot just needs to display image URL previews as big as images.

Doing this over the approach in this MSC has a few advantages in my opinion:

 1. No new API required. No additional Complexity is needed for servers and clients to implement.
 2. Saves storage Space. While the server still needs to proxy the image for privacy reasons, the server does not need to store the image forever. It can cache it short term and re-download it again later on demand.
 3. It requires no changes on the sending clients side, so it instantly works for everyone's messages. It also works in any message history predating this feature.
 4. It preserves the context that this is an image that's not created by the user, but downloaded from some URL.

Comment (uhoreg) [reply proposals/2370-resolve-url.md]: The semantics are different.  If you post a URL, then you are dependent on that server hosting the image.  Which means:

- if the server goes down, then you no longer have access to the image
- if the image on the server changes, then people will see a different image.  This may or may not be what you want, depending on context.

Yes, pasting a URL and relying on URL preview may work in some cases, but may not work in others.  For those cases where pasting a URL works, people are still free to use that.

Comment (NotAFile) [reply proposals/2370-resolve-url.md]: Well, I think those objections are correct, but they exist identically for all URL previews, not just images.

So, if those semantics are desired, why not still just reuse the URL preview API but query it before sending? URL previews obviously already support returning images.

So I understand why you'd want to get the image on the sender side now, but still not why we need a new, more limited version of an existing API for this functionality.

Comment (uhoreg) [reply proposals/2370-resolve-url.md]: If you just use the URL preview API:
- you don't know if the "preview" is exactly the same image, or a scaled/stripped version
- you have to download and re-upload the image, rather than just have the server download the image and keep it
- URL preview works that way for images, but what about audio or video files?

Comment (uhoreg) [reply proposals/2370-resolve-url.md:6]: > So, my suggestion is to expand it to audio and video files, as Telegram does.

"Previewing" a video file is semantically different from serving a copy of the file.  A preview of a video file could legitimately be a single frame of the video, which would be different from what this MSC proposes.  Similarly, a preview of a PDF would probably be a low-resolution image of the first page of the PDF.  And, as I mentioned, a preview for an image file could be a scaled-down version of the image, rather than a copy of the original image.

Basically, this API is intended for a different use from the preview API.
